gh-80789: Implement build-time pip bundling in ``ensurepip``

webknjaz opened the pull request:
Prior to this patch, Pip wheels were stored in the Git repository of
CPython. Git is optimized for text but these artifacts are binary. So
the unpleasant side effect of doing this is that the bare Git
repository size is being increased by the zip archive side every time
it is added, removed or modified. It's time to put a stop to this.

The patch implements an `ensurepip.bundle` module that is meant to be
called through `runpy` to download the Pip wheel and place it into the
same location as before. It removes the wheel file from the Git
repository and prevents re-adding it by defining a new `.gitignore`
configuration file.

The idea is that the builders of CPython are supposed to invoke the
following command during the build time:

```console
$ python -m ensurepip.bundle
```

This command will verify the existing wheel's SHA-256 hash and, if it
does not match, or doesn't exist, it will proceed to download the
artifact from PyPI. It will confirm its SHA-256 hash before placing it
into the `Lib/ensurepip/_bundled/` directory.

The only change concerning the Pip maintainers submitting pip version
bumps to the CPython repo is that they will no longer need to check in
the wheel file binary, but would instead have to update SHA-256 and pip
version strings at the end of `Lib/ensurepip/_structs.py`. Previously,
the pip version used to be hard-coded at the top of
`Lib/ensurepip/__init__.py` and SHA-256 was not included.

Every single line added or modified as a part of this change is also
covered with tests. Every new module has 100% coverage. The only
uncovered lines under `Lib/ensurepip/` are the ones that are
absolutely unrelated to this effort.

Resolves #80789.

Ref: https://bugs.python.org/issue36608.

## Coverage
```console
$ ./cpython-venv/bin/python -m coverage run --pylib --source=ensurepip Lib/test/test_ensurepip.py
...................................
----------------------------------------------------------------------
Ran 35 tests in 0.048s

OK
(cpython-pyenv-py3.11.4) 

$ ./cpython-venv/bin/python -m coverage report --show-missing --skip-covered
Name                        Stmts   Miss Branch BrPart  Cover   Missing
-----------------------------------------------------------------------
Lib/ensurepip/__init__.py      79      5     28      0    93%   21-39
Lib/ensurepip/__main__.py       2      2      0      0     0%   1-2
-----------------------------------------------------------------------
TOTAL                         226      7     77      0    97%

5 files skipped due to complete coverage.
```

<!-- issue-number: [bpo-36608](https://bugs.python.org/issue36608) -->
https://bugs.python.org/issue36608
<!-- /issue-number -->


<!-- gh-issue-number: gh-80789 -->
* Issue: gh-80789
<!-- /gh-issue-number -->

webknjaz committed:
👷Implement build-time pip bundling @ `ensurepip`

Prior to this patch, Pip wheels were stored in the Git repository of
CPython. Git is optimized for text but these artifacts are binary. So
the unpleasant side effect of doing this is that the bare Git
repository size is being increased by the zip archive side every time
it is added, removed or modified. It's time to put a stop to this.

The patch implements an `ensurepip.bundle` module that is meant to be
called through `runpy` to download the Pip wheel and place it into the
same location as before. It removes the wheel file from the Git
repository and prevents re-adding it by defining a new `.gitignore`
configuration file.

The idea is that the builders of CPython are supposed to invoke the
following command during the build time:

```console
$ python -m ensurepip.bundle
```

This command will verify the existing wheel's SHA-256 hash and, if it
does not match, or doesn't exist, it will proceed to download the
artifact from PyPI. It will confirm its SHA-256 hash before placing it
into the `Lib/ensurepip/_bundled/` directory.

Every single line added or modified as a part of this change is also
covered with tests. Every new module has 100% coverage. The only
uncovered lines under `Lib/ensurepip/` are the ones that are
absolutely unrelated to this effort.

Resolves #80789.

Ref: https://bugs.python.org/issue36608.

webknjaz commented on Makefile.pre.in:
```diff
@@ -2759,6 +2760,12 @@ patchcheck: all
+.PHONY: bundle-ensurepip-dists
+bundle-ensurepip-dists:
+	if test "x$(ENSUREPIP)" != "xno"  ; then \
+		$(RUNSHARED) $(PYTHON_FOR_BUILD) -m ensurepip.bundle -v ; \
```
:exclamation: I need help with this — my attempts to integrate the bundler into the build process fail in some CI jobs (supposedly because of the `./configure` CLI flags).

AA-Turner commented on .github/workflows/reusable-docs.yml:
```diff
@@ -100,7 +100,7 @@ jobs:
       run: ./configure --with-pydebug
     - name: 'Build CPython'
-      run: make -j4
+      run: make all-with-ensurepip-dists-bundled -j4
```
Do we need ensurepip for the doctests?

webknjaz replied:
Not sure. I thought something was failing, but that was at the time when I was trying to figure out the Makefile integration, and it didn't work. So maybe not, this needs some experimentation.

AA-Turner commented on Makefile.pre.in:
```diff
@@ -617,6 +617,9 @@ all:		@DEF_MAKE_ALL_RULE@
 		gdbhooks Programs/_testembed scripts checksharedmods rundsymutil
 
+.PHONY: all-with-ensurepip-dists-bundled
+all-with-ensurepip-dists-bundled: build_all bundle-ensurepip-dists
```
The name is quite long, and I'm unsure what 'dists' means here:
```suggestion
.PHONY: all-with-bundled-ensurepip
all-with-bundled-ensurepip: build_all bundle-ensurepip
```

webknjaz replied:
> I'm unsure what 'dists' means

Basically, this refers to Python distribution packages. As in wheel zipfiles or sdist tarballs.
Historically, `ensurepip` shipped with two wheels — `pip` and `setuptools`. Some time ago, Pradyun removed the `setuptools` one, but the infrastructure for having an iterable of dists remained in the source code. Hence the plural form.

Also, this make target doesn't control whether the ensurepip module is included by itself. It only downloads the wheel into the in-module location. So I think implying that this bundles ensurepip in the make target name is inaccurate.

When Pradyun and I last discussed it, we agreed that the list of things is probably no longer needed. So I was going to have a follow-up PR to replace it with a singular var at some point. I didn't want to include refactoring in this patch — it would make it harder to manage a non-atomic PR. But I suppose that follow-up could change the target name to specifically say “pip” instead of “dists”.

FWIW, I had problems embedding this into the Makefile. So maybe the RMs would have other preferences.

webknjaz replied:
I chose to make use of another make target name suggestion. See: https://github.com/python/cpython/pull/12791#discussion_r1307596457. Marking this thread as resolved.

AA-Turner commented on Makefile.pre.in:
```diff
@@ -2759,6 +2762,14 @@ patchcheck: all
 	$(RUNSHARED) ./$(BUILDPYTHON) $(srcdir)/Tools/build/stable_abi.py --all $(srcdir)/Misc/stable_abi.toml
 
+.PHONY: bundle-ensurepip-dists
+bundle-ensurepip-dists: $(BUILDPYTHON) platform
```
```suggestion
.PHONY: bundle-ensurepip
bundle-ensurepip: $(BUILDPYTHON) platform
```

webknjaz replied:
See https://github.com/python/cpython/pull/12791#discussion_r1307351594 for why this feels incorrect.

webknjaz replied:
I chose to make use of another make target name suggestion. See: https://github.com/python/cpython/pull/12791#discussion_r1307596457. Marking this thread as resolved.

AA-Turner reviewed (commented):
Brief initial look, I haven't gone over the changes to ``ensurepip`` fully.

I did want to ask on intended use scenarios here, though -- from Pradyun & Serhiy's comments on the issue, it seemed the goal was to have the release / build process pull in the latest version of ``pip`` (similar-ish to how other dependencies are vendored).

This PR though seems to implement a new ``ensurepip.bundler`` mechanism, which exposes a new module and has unclear semantics--is this for end users, redistributors, etc? If possible, perhaps we could either make the ``_bundler`` module private, or move the bundling mechanisms to ``Tools/`` for inclusion in to the build process.

Let me know if you'd like a more detailed review into the changes in ``ensurepip`` itself, I wanted to hold off for now as of these questions on the design level.

Thanks,
Adam

webknjaz committed:
Resolve file path @ `test_get_packages_with_dir`

webknjaz committed:
Download ensurepip pip dist @ ASAN GHA job

webknjaz commented on .github/workflows/reusable-docs.yml:
```diff
@@ -100,7 +100,7 @@ jobs:
-      run: make -j4
+      run: make all-with-ensurepip-dists-bundled -j4
     - name: 'Install build dependencies'
       run: make -C Doc/ PYTHON=../python venv
```
@AA-Turner so I looked into whether bundling pip into ensurepip is needed in doctests and the answer is “yes”. This line calls `../python -m venv` (so it uses the Python build under test and not some OS-provided copy). The `venv` module relies on `ensurepip` to provision its `site-packages/`. When there's no pip wheel, it crashes (as it should!).

webknjaz committed:
Run ensurepip bundle generation @ Win on GHA

webknjaz committed:
Expose TLS trust store to ASAN build @ GHA

webknjaz committed:
Call `ensurepip.bundle` @ win build script

webknjaz committed:
Invoke ensurepip.bundle via built Python @ Win

webknjaz committed:
Depend on dynamic default target @ make

webknjaz committed:
Update the equality check with `==` @ Win build

webknjaz committed:
Revert "Run ensurepip bundle generation @ Win on GHA"

This reverts commit 5bf0cdaaa0ff9ed12746710c44f14a69803a6d73.

webknjaz commented on Lib/ensurepip/_structs.py:
```diff
@@ -0,0 +1,251 @@
+
+PIP_REMOTE_DIST = _RemoteDistributionPackage(
+    'pip',
+    '23.2.1',
+    '7ccf472345f20d35bdc9d1841ff5f313260c2c33fe417f48c30ac46cccabf5be',
```
@pfmoore this is where the version information has moved to. I was considering moving it into a config file (like toml?) but that's something I wouldn't want to couple with this PR, it's big as it is.

pfmoore replied:
Thanks. So the pip release manager would need to calculate that sha256 value? That should be OK, although while I have tools on Windows to do this, I don't know if everyone does (I assume Unix has a command for this, although I've never checked).

(Actually, I just saw that you mentioned it can be copied from PyPI, so that's sufficient).

AA-Turner replied:
``sha256sum`` is usable on Unix (& WSL), or ``openssl sha256`` on both, or ``hashlib.sha256``!

webknjaz commented on PCbuild/build.bat:
```diff
@@ -175,6 +175,15 @@ echo on
+set suffix=
+if "%conf%"=="Debug" set suffix=_d
+set built_python_exe=%prefix%\python%suffix%.exe
+call "%built_python_exe%" -m ensurepip.bundle -v
```
I know this is ugly, but I don't know much about Windows :(
Improvement suggestions are welcome!

zware commented on PCbuild/build.bat:
```diff
@@ -175,6 +175,15 @@ echo on
+set suffix=
+if "%conf%"=="Debug" set suffix=_d
+set built_python_exe=%prefix%\python%suffix%.exe
+call "%built_python_exe%" -m ensurepip.bundle -v
```
There's a handy `"%dir%\..\python.bat"` that you can use instead.

webknjaz replied:
Is it guaranteed to locate the built interpreter version? I was a bit lost and haven't found how this file is generated to check it myself..

zware replied:
Yes; see https://github.com/python/cpython/blob/4116592b6f014a2720e9b09e2c8dec4bf4b4cd8f/PCbuild/python.vcxproj#L134-L150

(Note that `%dir%\..\python.bat` and `%dir%\find_python.bat` are very different :).  You want the former, which is also used in the PGO section several lines up.)

webknjaz replied:
I applied the suggested change. Marking this thread as resolved.

webknjaz replied:
@zware could you take a look at https://dev.azure.com/Python/cpython/_build/results?buildId=136036&view=logs&j=91c152bd-7320-5194-b252-1404e56e2478&t=c7e99cd8-4756-5292-d34b-246ff5fc615f&l=764 — any idea why it might be failing? I don't see how it might be related to the PR, but OTOH, it doesn't crash on `main`…

zware replied:
That particular build is apparently trying to run an `arm64` build on what is apparently an `amd64` runner, which isn't going to work out well :)

zware commented on Makefile.pre.in:
```diff
@@ -2759,6 +2762,14 @@ patchcheck: all
 	$(RUNSHARED) ./$(BUILDPYTHON) $(srcdir)/Tools/build/stable_abi.py --all $(srcdir)/Misc/stable_abi.toml
 
+.PHONY: bundle-ensurepip-dists
+bundle-ensurepip-dists: $(BUILDPYTHON) platform
```
ISTM this target name should explicitly call out that it is (probably) downloading something from the internet, something more along the lines of `download-ensurepip-blobs`.

webknjaz replied:
Fair point. I like this suggestion over https://github.com/python/cpython/pull/12791#discussion_r1306868995.

webknjaz replied:
I've applied this change, so I'm marking the thread as resolved.

webknjaz committed:
Use case-insensitive comparison @ win build script

webknjaz committed:
In download pip into ensurepip @ Azure Pipelines

zware reviewed (commented):
I am very much on board with the stated goal of this PR of removing the binary blobs from the repository.  I'm not fully convinced by the approach, though: this adds quite a lot of code to download (currently, and probably never again more than) one blob, and it took me several minutes of searching to find where the actual download was happening, which turns out to be buried in a module called `_structs`.

ISTM that if we're reaching out to the internet anyway, we could have a much lower maintenance burden here by making use of `bootstrap.pypa.io/get-pip.py` which could easily grow a `--download-only` option with a ~3 line implementation.  `get-pip.py --no-setuptools --no-wheel --download-only 'pip=={pip_version}' -w Lib/ensurepip/_bundled` would pretty much do everything we want as far as I can tell.  If we used a requirements file with it instead of the versioning argument, we could even have hash verification.

webknjaz committed:
Rename make target to `download-ensurepip-blobs`

Per suggestion at
https://github.com/python/cpython/pull/12791#discussion_r1307596457.

Co-Authored-By: Zachary Ware <zach@python.org>

webknjaz committed:
Call built Python via `python.bat` @ win build

Suggested via
https://github.com/python/cpython/pull/12791#discussion_r1307596328.

Co-Authored-By: Zachary Ware <zach@python.org>

webknjaz commented on Lib/ensurepip/_structs.py:
```diff
@@ -0,0 +1,251 @@
+        )
+
+        try:
+            first_matching_dist_wheel = sorted(dist_matching_wheels)[0]
```
Self note: the original code actually goes for the last matching:
```suggestion
            last_matching_dist_wheel = sorted(dist_matching_wheels)[-1]
```
